main.cpp

```diff
@@ -44,28 +44,376 @@ Purpose:  This project continues developing Project3.
+        unsigned int outputResistance {8};
+        float amplificationFactor {10.0f};
+        float thdNoise;
+        float voltage;
+
+        void increaseAmplification(float ampFactor);
+        void switchOutResistance(unsigned int outRes);
+        void switchInResistance(unsigned int inRes);
+    };
+
+    Preamp preA;
+
+    void adjustVolume(Preamp pre);
+    bool switchOn(); 
+    bool changeInp(char inputType); 
+    void ejectCd(int time);
+}; 
+
+Amplifier::Amplifier() : inputs(8), wattage(45), maxVolume(10.0)
+{
+    std::cout << "Amplifier created!" << '\n';
+}
+
+Amplifier::~Amplifier()
+{
+    std::cout << "Amplifier Object destruceted!" << '\n';
+}
+
+Amplifier::Preamp::Preamp() : thdNoise(0.0001f), voltage(12)
+{
+    std::cout << "Preamp created!" << '\n';
+}
+
+void Amplifier::Preamp::increaseAmplification(float ampFac)
+{
+    std::cout << "Factor set: " << ampFac << '\n';
+}
+
+void Amplifier::Preamp::switchOutResistance(unsigned int outR)
+{
+    std::cout << "Output Resistance set: " << outR<< '\n';
+}
+
+void Amplifier::Preamp::switchInResistance(unsigned int inR)
+{
+    std::cout << "Input Resistance set: " << inR<< '\n';
+}
+
+void Amplifier::adjustVolume(Preamp pre)
+{
+    pre.increaseAmplification(5);
+}
+
+bool Amplifier::switchOn()
+{
+    return true;
+}
+
+bool Amplifier::changeInp(char type)
+{
+    switch(type) 
+    {
+        case 'A' :
+            std::cout << "Class A" << '\n'; 
+            break;
+        case 'B' :
+            std::cout << "Class B" << '\n';
+            break;
+        default :
+            std::cout << "Invalid type" << '\n';
+   }
+
+    return (type == ('A' | 'B') ? true : false);
+}
+
+void Amplifier::ejectCd(int time)
+{
+    auto i {0};
+    while(i != time)
+    {
+        ++i;
+    }
+    std::cout << "Ejecting CD took " << i << " seconds" << '\n';
+}
+
 /*
  new UDT 4:
  with 2 member functions
  */
 
+struct Kitchen
+{
+    Kitchen();
+    ~Kitchen();
+
+    Fridge fridge;
+    CoffeeMachine machine;
+    bool table {false}; // true if something is on the table
+
+    void makeCappucino();
+    void fillCoffeeTank();
+
+};
+
+Kitchen::Kitchen() 
+{
+    std::cout << "Kitchen Object constructed!" << '\n';
+}
+
+Kitchen::~Kitchen() 
+{
+    std::cout << "Kitchen Object destructed!" << '\n';
+}
+
+void Kitchen::makeCappucino()
+{
+    bool opened = this->fridge.openDoor(this->fridge.door1, 45);
+
+    if(opened == true)
+    {
+        this->machine.heatUp(10);
+        this->machine.foamMilk(10);
+        this->machine.makeCoffee(10, machine.crushBeans());
+    }
+
+    std::cout << "Done, making cappucino!" << '\n';
+}
+
+void Kitchen::fillCoffeeTank()
+{
+    this->machine.tankVolume = fridge.streamWater(this->fridge.waterTankVolume);
+
+    std::cout << "Coffee tank filled with water!" << '\n';
+}
+
+
 /*
  new UDT 5:
  with 2 member functions
  */
 
+struct KitchenRoom
+{
+    KitchenRoom();
+    ~KitchenRoom();
+
+    Kitchen kitchen;
+    Amplifier radio;
```

Comment:
How is a kitchen room different from kitchen?

Come up with another UDT. 
- [ ] revise